proposals/0006-Keyboard-Friendly-Virtualization.md

```diff
@@ -0,0 +1,179 @@
+---
+title: "RFC0006: Keyboard-friendly List Virtualization Behavior"
+author: Nick Gerleman
+date: February 2021
+---
+
+# RFC0006: Keyboard-friendly List Virtualization Behavior
+
+## Summary
+
+VirtualizedList virtualizes away items based on proximity to the lists viewport,
+without consideration for other inputs such as focus or selection. This
+strategy limits the ability to correctly implement some common keyboard
+interactions. This especially impacts desktop platforms (react-native-web,
+react-native-windows, react-native-macos), but also impacts the tablet with
+keyboard form-factor on iPadOS/Android.
+
+This proposal outlines these issues, proposing a set of behavior changes and new
+APIs to address them. It does not address concrete implementation.
+
+## Basic example
+
+```jsx
+<FlatList
+  {...otherProps}
+  keepAreasRealized={{
+    focused: false,
+    home: [true, {windowSize: 1}],
+  }}
+/>
+```
+
+## Motivation
+
+Keyboard users expect to be able to navigate between between views inside of a
+list via tab-loop or arrow keys. A component with focus may move out of
+visibility by panning or scrolling a parent surface. The focused component may
+be virtualized away after leaving visibility.
+
+The removal of this native component disrupts state needed to move focus/selection.
+
+## Detailed design
+
+### Realization windows
+
+Focus may be moved by the native UI framework without waiting on a new component to be rendered. This forces us to
+prevent virtualization (keep realized) any components which focus may be moved
+to. We may see additional navigation after the initial focus change, unblocked
+on the rendering of new components. This necessitates a chain of additional
+realized components.
+
+Common keyboarding scenarios should ideally work by default, without having to change
+settings. Some of these scenarios include
+
+1. Navigating from one child component of a rendered item to another
+1. Arrowing between focused items
+1. Shifting focus from currently focused item via page up/down
+1. Shifting focus to first/last item via home/end keys
+
+`VirtualizedList` has enough internal knowledge to transparently support these
+scenarios. This can be accomplished by adding three **"realization windows"**:
+1. **"focused"** Realize cells around a cell with active focus. Can be
+tracked by listening to bubbling onFocus in the default `CellRendererComponent`.
```

Comment:
Maybe it would make sense to listen to capture phase instead of bubbling phase, in case something inside the cell wants to stop upward propagation.

Reply:
I agree this makes sense to track in the capture phase.